### DISCUSSION TITLE
NPC id and index lifetime

### DISCUSSION THREAD

**Discussion by anttikarhu** [runelite/runelite#16140] [Q&A] [answered]
Hi!

I was wondering, what is the lifetime of the NPC id and index attributes? I’m trying to identify single NPC instances (like a Goblin which spawns at Lumby) over a longer time period (indefinetly if possible), but I’m unsure if those two numbers would be enough?

They seem to remain the same based on my short testing of a workday but I’m not sure how they behave in the long run.

Is there any other way to identify an individual NPC you could think of?

**Comment by Nightfirecat** [comment, marked as answer]
NPC ID will not change over time, unless Jagex updates the NPC definition itself (in which case, you would see [a corresponding NpcID update on our side after said release](https://github.com/runelite/runelite/commit/fe2cba95a4222fdd6d09cf72b0813f70e232e10d)). An NPC's index is generally valid for as long as the servers are not restarted. I would not rely on it being the same between client sessions. [You can reference the NPC Indicators plugin's code](https://github.com/runelite/runelite/tree/master/runelite-client/src/main/java/net/runelite/client/plugins/npchighlight) to see how we handle tracking a specific NPC's spawn location and duration. I'd expect whatever testing you're doing could use much of the same code.

**Reply by anttikarhu** [reply]
Awesome, I will check that plugin out. Thank you for the information!

**Reply by anttikarhu** [reply]
For what I could see from the npcindicator plugin, the npcs are identified by the id+index, but the server boot doesen’t matter in that use case if the index changes.

thank you for your help!
